src/librbd/ExclusiveLock.cc

```diff
@@ -134,11 +134,13 @@ void ExclusiveLock<I>::shut_down(Context *on_shut_down) {
   {
```

Comment:
If you wrap the ```on_shut_down``` around a ```C_Gather```, you could pass a new gather subscription down to ```execute_action``` and then activate the gather after you handle the peer notification to ensure the callback is properly delayed.